build-logic/build.gradle.kts

```diff
@@ -6,14 +8,29 @@ repositories {
     mavenCentral()
 }
 
+kotlin {
+    jvmToolchain(20)
+}
+
+// `build-logic` can produce Java 20 compatible code, but only after the plugin is built with kotlin 1.9,
```

Comment:
I'll answer here first, and will later create a separate issue: 

>It is not clear to me what should be done and why it should be done.

Project `/build-logic` that holds all Gradle Plugins definitions produces `jar` file that is later appended to `/` build classpath. 

We know the `/` is built with Java 20, so it would be fine `/build-logic` if produced java-20 compatible bytecode - we know  `/` project is able to consume such.
Unfortunately, due to https://youtrack.jetbrains.com/issue/KT-57669/Add-Java-20-to-JvmTarget, `/build-logic` project can't be compiled with Java 20 target yet (JVM target => produced bytecode compatibility). The fix is available in Kotlin Gradle Plugin 1.9, but because `/build-logic` and `/` share a common Version Catalog  (`.toml`) it can't yet be upgraded and both projects use lower Kotlin version. 

Theoretically, due to the fact these are 2 separated projects, `/build-logic` could use Kotlin 1.9, while the `/` still uses Kotlin 1.8.22, but such change is not worth the extra complexity.

Regarding **what** should be done? After `/build-logic` has Kotlin Gradle Plugin 1.9.0 applied, there are 2 (well 3) options:
1. Setting explicit JVM target version here could be completely skipped, and its value would fully rely on the implicit behavior that toolchain JDK = target JVM. (Point 1 [here](https://docs.gradle.org/8.2/userguide/building_java_projects.html#sec:java_cross_compilation)). tl;dr remove the extra code I added, leave only `jvmToolchain(20)` setting.
2. Keep setting explicit JVM target version, but upgrade produced bytecode from Java 17 to Java 20
3. Keep the current config, it's also _fine_, but the KGP won't be able to use new features available in higher Java version when producing JVM bytecode. 

>To what plugin are you referring?

I was referring to all plugins published from `/build-logic` project:
![image](https://github.com/pinterest/ktlint/assets/36954793/c3fafbd7-751d-4b5c-bee0-42b4a4ae1357)

Would a
```suggestion
// it would be OK if `build-logic` produced Java 20 compatible bytecode, but this requires using Kotlin Gradle Plugin 1.9, which can't be used yet. visit {{ link to the issue I'll create later }} for more details
```
explain my intentions better? 



Please let me know if anything I'm saying is not clear! 
My main point here is I wanted to stress out this repository consists of 2 separate builds/root projects: 
a) a `/build-logic` - fully internal, produces Gradle Plugins, used to configure `/` project 
b) `/` containing ktlint library, uses `a)` to configure itself, but doesn't expose a) dependencies to ktlint consumers. 
The fact there are 2 separate builds makes the amount of configuration code feel duplicated :)

Reply:
Clear. And yes, the updated comment is much more clear. Now, I realized that it even can be written as a TODO:
```
TODO: After upgrade to Kotlin Gradle Plugin 1.9, ....
```
Before each release, I check all TODO's and if a constraint is met, I will always try to actually resolve before releasing.